src/MIDebugEngine/Engine.Impl/Disassembly.cs

```diff
@@ -166,10 +143,123 @@ public Disassembly(DebuggedProcess process)
                 }
                 var kv = block.TryFetch(address, nInstructions, out ret);
             }
-
             return ret;
         }
 
+        /// <summary>
+        /// Seek backward n instructions from a target address
+        /// </summary>
+        /// <param name="address">target address</param>
+        /// <param name="nInstructions">number of instructions</param>
+        /// <returns> address - n on failure, else the address of an instruction n back from the target address</returns>
+        public async Task<ulong> SeekBack(ulong address, int nInstructions)
+        {
+            ICollection<DisasmInstruction> ret = null;
+            ulong defaultAddr = address >= (ulong)nInstructions ? address - (ulong)nInstructions : 0;
+
+            lock (_disassemlyCache)
+            {
+                // check the cache
+                var kv = _disassemlyCache.FirstOrDefault((p) => p.Value.TryFetch(address, -nInstructions, out ret));
+                if (kv.Value != null)
+                    return ret.First().Addr;
+            }
+            ulong endAddress;
+            ulong startAddress;
+            var range = await _process.FindValidMemoryRange(address, (uint)(_process.MaxInstructionSize * nInstructions)+1, (int)(_process.MaxInstructionSize * -nInstructions));
+            startAddress = range.Item1;
+            endAddress = range.Item2;
+            if (endAddress - startAddress == 0) // bad address range, no instructions
+            {
+                return defaultAddr;
+            }
+            lock (_disassemlyCache)
+            {
+                // check the cache with the adjusted range
+                var kv = _disassemlyCache.FirstOrDefault((p) => p.Value.TryFetch(startAddress, address < endAddress ? address : endAddress, out ret));
+            }
+            if (ret == null)
+            {
+                DisasmInstruction[] instructions = await Disassemble(_process, startAddress, endAddress);
```

Comment:
Gdb seems to do a good job of finding the correct instructions. That said, we always call VerifyDisassembly for any newly disassembled address range. If we fail to find the "right" address as the start of an instruction - because it is not really the start of an instruction - then we just use whatever gdb first returned as disassembly for the address range.